src/encoding/hybrid_rle/decoder.rs

```diff
@@ -3,38 +3,75 @@ use super::{super::ceil8, HybridEncoded};
 
 /// An iterator that, given a slice of bytes, returns `HybridEncoded`
 pub struct Decoder<'a> {
-    values: &'a [u8],
-    num_bits: u32,
+    inner: std::iter::Flatten<HybridIterator<'a>>,
 }
 
 impl<'a> Decoder<'a> {
-    pub fn new(values: &'a [u8], num_bits: u32) -> Self {
-        Self { values, num_bits }
+    pub fn new(values: &'a [u8], num_bits: u32, length: usize) -> Self {
+        Self {
+            inner: HybridIterator {
+                values,
+                num_bits,
+                remaining: length,
+            }
+            .flatten(),
+        }
     }
 }
 
 impl<'a> Iterator for Decoder<'a> {
-    type Item = HybridEncoded<'a>;
+    type Item = u32;
 
     fn next(&mut self) -> Option<Self::Item> {
+        self.inner.next()
+    }
+}
+
+pub struct HybridIterator<'a> {
+    values: &'a [u8],
+    num_bits: u32,
+    remaining: usize,
+}
+
+impl<'a> Iterator for HybridIterator<'a> {
+    type Item = HybridEncoded<'a>;
+
+    fn next(&mut self) -> Option<HybridEncoded<'a>> {
         if self.values.is_empty() {
             return None;
         }
         let (indicator, consumed) = uleb128::decode(self.values);
         self.values = &self.values[consumed..];
         if indicator & 1 == 1 {
             // is bitpacking
-            let bytes = (indicator as usize >> 1) * self.num_bits as usize;
-            let result = Some(HybridEncoded::Bitpacked(&self.values[..bytes]));
-            self.values = &self.values[bytes..];
+            let num_bits = self.num_bits as usize;
+            let num_bytes = (indicator as usize >> 1) * num_bits;
+            let length = (indicator as usize >> 1) * 8;
+            let run_length = std::cmp::min(length, self.remaining);
+            let result = Some(HybridEncoded::Bitpacked {
+                compressed: &self.values[..num_bytes],
+                num_bits,
+                run_length,
+            });
+            self.remaining -= run_length;
+            self.values = &self.values[num_bytes..];
             result
         } else {
             // is rle
             let run_length = indicator as usize >> 1;
             // repeated-value := value that is repeated, using a fixed-width of round-up-to-next-byte(bit-width)
             let rle_bytes = ceil8(self.num_bits as usize);
-            let result = Some(HybridEncoded::Rle(&self.values[..rle_bytes], run_length));
+
+            let pack = &self.values[0..rle_bytes];
+            let mut value_bytes = [0u8; std::mem::size_of::<u32>()];
+            pack.iter()
+                .enumerate()
+                .for_each(|(i, byte)| value_bytes[i] = *byte);
+            let value = u32::from_le_bytes(value_bytes);
```

Comment:
> What do you think about keeping this decoder as is, and instead have `rle_decode` implemented as returning an iterator? It may solve both of our use-cases?

That sounds like it should work, seems all users call `into_iter` on the rle decoded vec anyway. It should then even be possible to expose the iterator directly in the api instead of first creating a `Vec<Option<Vec<u8>>>` for example, whose contents then have to be copied again into the arrow format.

Reply:
Now that I realized that arrow2 is using the lower level api I understand what you actually meant :)
Do you think the code in arrow2 could be simplified by exposing an iterator api in parquet2, without loosing performance?

Reply:
So, we already have an iterator API, that returns either variant, RLE or Bitpacked. This is made so that arrow2 can use it without having to incur the cost of bit-unpacking the bitpacked to then compare an `i32` to 1 (max def level).

There will be a performance degradation if we move to iterators over i32, which is why I was trying to suggest to keep the encoder as is, and offer a higher-end iterator to iterate over it with `i32` (the `level.rs`).